aclk/agent_cloud_link.c

```diff
@@ -1336,24 +1336,31 @@ void *aclk_main(void *ptr)
     struct netdata_static_thread *static_thread = (struct netdata_static_thread *)ptr;
     struct netdata_static_thread *query_thread;
 
-    if (!netdata_cloud_setting) {
-        info("Killing ACLK thread -> cloud functionality has been disabled");
-        static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;
-        return NULL;
-    }
+#if defined( DISABLE_CLOUD ) || !defined( ENABLE_ACLK)
+    info("Killing ACLK thread -> cloud functionality has been disabled");
+    static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;
+    return NULL;
+#endif
 
     info("Waiting for netdata to be ready");
     while (!netdata_ready) {
         sleep_usec(USEC_PER_MS * 300);
     }
 
+    while (!netdata_cloud_setting) {
+        sleep_usec(USEC_PER_SEC * 5);
+    }
+
```

Comment:
When cloud is not enabled and it enters the loop to wait, hitting control -c will delay shutdown with the following message:

```
Master thread ACLK_Main takes too long to exit. Giving up..
``` 

this is because the ACLK_Main is canceled by `cancel_threads()` but then it waits for up to 5 seconds for the thread to change the `thread->enabled` to `NETDATA_MAIN_THREAD_EXITED` which doesn't happen.

We may need to use `netdata_thread_disable_cancelability()` earlier to handle this because there are sections of the code following that loop that can suffer from the same problem is control - C is hit.

Another option might be to extend the thread state and add `NETDATA_MAIN_THREAD_RUNNING_NO_CANCEL` and have `cancel_threads()` skip those threads that will rely on `netdata_exit` to set the status to `NETDATA_MAIN_THREAD_EXITED` and avoid the 5 second delay during shutdown.

Reply:
This opens an ugly can of worms. I'm changing this but it affects several other points in the code, new loop for the disabled case will be:
```
    while (!netdata_cloud_setting) {
        sleep_usec(USEC_PER_SEC * 1);
        if (netdata_exit) {
            static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;
            return NULL;
        }
    }
```
But the code in `cancel_main_threads()` seems to be wrong (@mfundul please comment). If a thread *is* cancelled on the first pass we don't check if the cancel was successful, which means we wait for the status to change but there is nothing running to change it. Moving the `netdata_disable_thread_cancelability()` earlier in the thread (as @stelfrag suggests) will hide the problem in this case (as we will flip the flag and exit) but which threads in the system are being cancelled at shut-down and how does their status get updated?

Also I can see two other problems:
1. The wait to be claimed loop has the same problem.
2. The `aclk_decode_base_url` is now in the wrong place as it can be updated at claiming time.

Reply:
Actually even with the change below you don't get a change to set  `static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;` because cancel has already been called on the thread (`netdata_exit` is set after that cancel-threads-loop) 


```
    while (!netdata_cloud_setting) {
         sleep_usec(USEC_PER_SEC * 1);
         if (netdata_exit) {
             static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;
             return NULL;
         }
     }
```

the ugly way to be able to check for `netdata_exit` is 

```
netdata_thread_disable_cancelability();
while (!netdata_cloud_setting) {
         sleep_usec(USEC_PER_SEC * 1);
         if (netdata_exit) {
             static_thread->enabled = NETDATA_MAIN_THREAD_EXITED;
             return NULL;
         }
}
netdata_thread_enable_cancelability();
```

Yes same problem with the claim loop. There is also memory cleanup issue when `goto exited` in that case (variables not allocated)

``` 
    freez(aclk_username);
    freez(aclk_password);
    freez(aclk_hostname);
    freez(aclk_port);
```

Reply:
`cancel_main_threads()` assumes that `pthread_cancel()` will always succeed, and while this may be wrong, in practice `pthread_cancel()` has never failed. If cancelability has been disabled then `pthread_cancel()` gets queued and succeeds anyway.

Reply:
There was a longer context that I skipped - we ripped out the cleanup handler in the ACLK thread because of the LWT changes. As a result, we need to enable cancel ability earlier and manually do cleanup in that particular thread - it is "special".

I've pushed a commit that cleans up the shutdown behaviour, could you both please check it and if it seems robust then I will merge this PR onto the "go-live" branch as I need to switch over to looking at something else (the renegotiation fix).

Reply:
I've tested shutdown in:
* The cloud disabled state
* The unclaimed state
* The claimed state
* The connected state
I can't see any problems although I forgot to run it through valgrind (c.f. @stelfrag comment's above about memory cleanup).